src/ngx_http_modsecurity_body_filter.c

```diff
@@ -135,47 +134,43 @@ ngx_http_modsecurity_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
     }
 #endif
 
+    int is_request_processed = 0;
     for (; chain != NULL; chain = chain->next)
     {
-/* XXX: chain->buf->last_buf || chain->buf->last_in_chain */
-        if (chain->buf->last_buf) {
-            buffer_fully_loadead = 1;
+        u_char *data = chain->buf->pos;
+        int ret;
+
+        msc_append_response_body(ctx->modsec_transaction, data, chain->buf->last - data);
```

Comment:
Let's use `ngx_buf_size()` here.

Reply:
`ngx_buf_size` is defined as
```
#define ngx_buf_size(b)                                                      \
    (ngx_buf_in_memory(b) ? (off_t) (b->last - b->pos):                      \
                            (b->file_last - b->file_pos))

```
If somehow we happend to receive a `buf` which is completely in a file (i.e., buf->in_file == true but ngx_buf_in_memory(buf) == false) then .pos = .last = 0 but .file_last - .file_post > 0, so we will try to dereference a non-zero length block at NULL pointer .pos. But with current `.last - .pos`  we would erroneously assume that we have no data but won't segfault at least.

But from the other hand I did a quick testing with gdb, even when a response body is buffered to a temporary file, nginx supplies a `buf` which has both `.in_file` and `.temporary` set (`ngx_buf_in_memory(buf) == true` =>  `.pos`, `.last` are valid).
ngx_http_charset_filter_module also references .pos/.last in `ngx_http_charset_recode_{to,from}_utf8` without worries . So I think it is guaranteed that response body filters get chain bufs with `ngx_buf_in_memory(buf) == true` ?

One more thing, I did a quick grep of Nginx sources and see that ngx_buf_size is not used that much, whereas `.last - .pos` is much more common... maybe we should stick to the latter?

Reply:
@turchanov well ok - let's keep `.last - .pos`.